man/man1/cru2clm.1
.TH cru2clm 1  "USER COMMANDS"
.SH NAME
cru2clm \- convert raw binary data files to clm data files for LPJmL
.SH SYNOPSIS
.B cru2clm
[\-h] [\-firstyear first] [\-lastyear last] [\-nyear n] [\-scale s] [\-firstcell f] [\-ncell n] [\-nbands n] [\-swap] [\-yearcell]
.I  crufile clmfile
.SH DESCRIPTION
Program converts raw binary data file consisting of 16bit integer values into a file format suitable for LPJmL. It is primarily intended to convert climate data files from the Climate Research Unit (CRU).
.SH OPTIONS
.TP
\-h
Print a short help text.
.TP
\-firstyear first
First year in CRU file (default is 1901).
.TP
\-lastyear last
Last year in CRU file.
.TP
\-nyear n
Number of years in CRU file (default is 103).
.TP
\-scale s
Set scale to s (default is 1).
.TP
\-firstcell f
First cell in data file (default 0).
.TP
-ncell  n
Number of cells in data file (default is 67420).
.TP
\-nbands n
Number of bands in data/output file (default is 12).
.TP
\-swap
Change byte order in CRU file.
.TP
\-yearcell
Does not revert order in CRU file.
.TP
.I crufile
Filename of binary CRU data file.
.TP
.I clmfile
Filename of clm data file written.
.SH EXAMPLES
.TP
Create clm data file from CRU dataset changing byte order:
.B cru2clm
\-swap tmp_1901-2003.bin tmp2003.clm
.PP
.SH EXIT STATUS
.B cru2clm
returns a zero exit status if the clm file is created.
Non zero is returned in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
lpjml(1), asc2clm(1), txt2clm(1), grid2clm(1), cft2clm(1), arr2clm(1), catclm(1), clm(5)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)

### Changed

- Axis names and missing values for NetCDF output are read from JSON file `netcdf.cjson`.
- Warnings in `addheader` replaced by terminating errors.
- Function `openinputfile()` returns an error if cell size differs in JSON metafile.
- Function `openmetafile()` returns an error if binary file is empty.
- `USE_NETCDF4` compile option replaced by `"netcdf4"` boolean flag in the LPJmL configuration file. If set compression of NetCDF4 files can be enabled.

### Added

- Environment variables `LPJNOPP` and `LPJPEDANTIC` added. If set to `true`, preprocessor is disabled and pedantic checking is enabled, respectively.
- Info about `--help` option added in case of error in command line arguments.
- Option `-config` added to `bin2cdf`, `country2cdf` and `clm2cdf` in order to read axis names and missing values from a JSON file. File `netcdf.json` added as a template.
- Option `-json` added to `arr2clm`, `asc2clm`, `txt2grid`, `txt2clm`, `getcountry`, and `cdf2coord` in order to create additional JSON metafiles.
- Option `-inpath` added to `regridlpj` in order to set directory where data can be found.
- Option `-double` added to `txt2grid` in order to create grid files with coordinate values of type double.
- Option `-timestep` added to `addheader`.
- Check for correct number of time steps added in `openclimate_netcdf.c`.
- Option `sum`, `tsum` and `tmean` to `mathclm` added.


## [5.9.9] - 2024-09-04

- author: Christoph Müller (cmueller@pik-potsdam.de)
- code review: Stephen Wirth (wirth@pik-potsdam.de), Fabian Stenzel (stenzel@pik-potsdam.de)

### Added

- new options `PRESCRIBED_SDATE_ALL_RAINFED` and `PRESCRIBED_SDATE_ALL_IRRIG` to allow using the same sowing dates based on the rainfed or irrigated seasons respectively
- new options `PRESCRIBED_CROP_PHU_ALL_RAINFED` and `PRESCRIBED_CROP_PHU_ALL_IRRIG` to allow using the same PHU requirements based on the rainfed or irrigated seasons respectively

### Changed

- refactored `crop_option_restart` to `crop_phu_option_restart` in struct `Config` and `crop_option` to `crop_phu_option` in struct `Restartheader` for greater clarity
- refactored options `"new"` to `"vbussel15"` and `"old"` to `"bondau07"` for `crop_phu_options` where `"vbussel15"` is an implementation based on [van Bussel et al. 2015](http://dx.doi.org/10.1111/geb.12351) 

### Removed

- removed unnecessary check for file of prescribed `sdates` in `fileexist.c`

### Fixed

- fixed missing initialization of pointer `map` in cft2bin.c that caused abortion of compilation with `-Werror` otherwise


## [5.9.8] - 2024-09-02

### Contributors

- author: Christoph Müller (cmueller@pik-potsdam.de)
- code review: Susanne Rolinski (rolinski@pik-potsdam.de), Fabian Stenzel (stenzel@pik-potsdam.de)

### Changed

- included latest reference in .zenodo.json for syncing github and gitlab repositories. Future transfer of code between gitlab and github should be smoother now as both are at the same commit history now


## [5.9.7] - 2024-08-30

### Contributors

- author: David Hoetten (davidho@pik-potsdam.de)
- code review: Sibyll Schaphoff (sibylls@pik-potsdam.de), Werner von Bloh (bloh@pik-potsdam.de)

### Changed

- order of freezing of different water soil water components changed: free water freezes last to simply mathematical description of model


## [5.9.6] - 2024-08-29

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: David Hoetten (davidho@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de)

### Added

- Utility `cdf2grid` added to convert grid files in NetCDF format into grid CLM files.
- Option `-json` and `-raw` added to `cdf2coord` utility to write additional JSON metafile and to write without CLM header.

### Changed

- Check added that dimension must be greater 1 for variable in NetCDF file in `cdf2coord.c`, `input_netcdf.c` and `coord_netcdf.c`.
- Datatype of longitude/latitude array in `coord_netcdf.c` changed from float to double for consistency, variable name correctly printed in error messages.
- Intel and clang compiler recognized in `printflags.c`.

### Fixed

- Test for `null` corrected for `"output"` in `fscanoutput.c`.
- Missing `break` added in `switch` statement in `receiver_coupler.c`.
- Code changed to compile successfully without `-DSAFE`, `-DUSE_NETCDF`, and `-DUDUNITS` compilation flag.
- Unit for `"estab_storage_n"` corrected to `"gN/m2"` in `outputvars.cjson`.
- Scaling set to 1 if datatype is not short in `cdf2coord.c` and `cdf2clm`.


## [5.9.5] - 2024-07-19

### Contributors

- author: Sebastian Ostberg (ostberg@pik-potsdam.de)
- code review: Christoph Mueller (cmueller@pik-potsdam.de), Werner von Bloh (bloh@pik-potsdam.de)

### Fixed

- Fix a regression for `NBP` output in `fwriteoutput` introduced in version 5.9.4


## [5.9.4] - 2024-07-18

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Jens Heinke (heinke@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de)

### Added

- Warning added if time step of output is longer than output interval.
- Cycle length of spinup and number of spinup years checked for positive values in `fscanconfig.c`.

src/utils/cru2clm.c
/**************************************************************************************/
/**                                                                                \n**/
/**                      c  r  u  2  c  l  m  .  c                                 \n**/
/**                                                                                \n**/
/**     Converts CRU data files into LPJ climate data files                        \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#undef USE_MPI /* no MPI */

#include "lpj.h"

#define NYEAR 103  /* default value for number of years */
#define FIRSTYEAR 1901 /* default value for first year */
#define FIRSTCELL 0
#define NCELL 67420
#define USAGE    "Usage: cru2clm [-h] [-firstyear first] [-lastyear last]\n"\
                 "       [-nyear n] [-scale s] [-firstcell f] [-ncell n] [-nbands nb] [-swap]\n"\
                 "       [-yearcell] crufile clmfile\n"

typedef short Data[NMONTH];

int main(int argc,char **argv)
{
  FILE *file;
  Data *data;
  char *endptr;
  Header header;
  long int k;
  int i,j,n;
  Bool swap;
  header.nyear=NYEAR;
  header.firstyear=FIRSTYEAR;
  header.order=CELLYEAR;
  header.firstcell=FIRSTCELL;
  header.ncell=NCELL;
  header.nbands=NMONTH;
  header.cellsize_lon=header.cellsize_lat=0.5;
  header.datatype=LPJ_SHORT;
  header.scalar=1;
  swap=FALSE;
  for(i=1;i<argc;i++)
    if(argv[i][0]=='-')
    {
      if(!strcmp(argv[i],"-h"))
      {
        printf("cru2clm (" __DATE__ ") - convert cru data files to\n"
               "       clm data files for LPJmL version " LPJ_VERSION "\n\n");
        printf(USAGE
               "\nArguments:\n"
               "-h               print this help text\n" 
               "-firstyear first first year in cru file (default is %d)\n"
               "-lastyear last   last year in cru file\n"
               "-nyear n         number of years in cru file (default is %d)\n"
               "-scale s         set scale to s, default is 1\n"
               "-firstcell       first cell in data file (default %d)\n"
               "-ncell           number of cells in data file (default is %d)\n"
               "-nbands          number of bands in data/output file (default is %d)\n"
               "-swap            change byte order in cru file\n"
               "-yearcell        does not revert order in cru file\n"
               "crufile          filename of cru data file\n"
               "clmfile          filename of clm data file\n\n"
               "(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file\n",
               FIRSTYEAR,NYEAR,FIRSTCELL,NCELL,NMONTH);
        return EXIT_SUCCESS;
      }
      else if(!strcmp(argv[i],"-nyear"))
      {
        if(i==argc-1)
        {
          fputs("Argument missing after '-nyear' option.\n"
                USAGE,stderr);
          return EXIT_FAILURE;
        }
        header.nyear=strtol(argv[++i],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid value '%s' for option '-nyear'.\n",
                  argv[i]);
          return EXIT_FAILURE;
        }
        if(header.nyear<1)
        {
          fprintf(stderr,"Number of years=%d must be greater than zero.\n",header.nyear);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[i],"-firstyear"))
      {
        if(i==argc-1)
        {
          fputs("Argument missing after '-firstyear' option.\n"
                USAGE,stderr);
          return EXIT_FAILURE;
        }
        header.firstyear=strtol(argv[++i],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid value '%s' for option '-firstyear'.\n",
                  argv[i]);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[i],"-lastyear"))
      {
        if(i==argc-1)
        {
          fputs("Argument missing after '-lastyear' option.\n"
                USAGE,stderr);
          return EXIT_FAILURE;
        }
        header.nyear=strtol(argv[++i],&endptr,10)-header.firstyear+1;
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid value '%s' for option '-lastyear'.\n",
                  argv[i]);
          return EXIT_FAILURE;
        }
        if(header.nyear<1)
        {
          fprintf(stderr,"Last year=%d less than first year=%d.\n",header.firstyear+header.nyear-1,header.firstyear);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[i],"-scale"))
      {
        if(i==argc-1)
        {
          fputs("Argument missing after '-scale' option.\n"
                USAGE,stderr);
          return EXIT_FAILURE;
        }
        header.scalar=(float)strtod(argv[++i],&endptr);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid value '%s' for option '-scale'.\n",
                  argv[i]);
          return EXIT_FAILURE;
        }
        if(header.scalar<=0)
        {
          fprintf(stderr,"Scalar=%g must be greater than zero.\n",header.scalar);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[i],"-firstcell"))
      {
        if(i==argc-1)
        {
          fputs("Argument missing after '-firstcell' option.\n"
                USAGE,stderr);
          return EXIT_FAILURE;
        }
        header.firstcell=strtol(argv[++i],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid value '%s' for option '-firstcell'.\n",
                  argv[i]);
          return EXIT_FAILURE;
        }
        if(header.firstcell<1)
        {
          fprintf(stderr,"First cell=%d must be greater than zero.\n",header.firstcell);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[i],"-ncell"))
      {
        if(i==argc-1)
        {
          fputs("Argument missing after '-ncell' option.\n"
                USAGE,stderr);
          return EXIT_FAILURE;
        }
        header.ncell=strtol(argv[++i],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid value '%s' for option '-ncell'.\n",
                  argv[i]);
          return EXIT_FAILURE;
        }
        if(header.ncell<1)
        {
          fprintf(stderr,"Number of cells=%d must be greater than zero.\n",header.ncell);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[i],"-nbands"))
      {
        if(i==argc-1)
        {
          fputs("Argument missing after '-nbands' option.\n"
                USAGE,stderr);
          return EXIT_FAILURE;
        }
        header.nbands=strtol(argv[++i],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid value '%s' for option '-nbands'.\n",
                  argv[i]);
          return EXIT_FAILURE;
        }
        if(header.nbands<1)
        {
          fprintf(stderr,"Number of bands=%d must be greater than zero.\n",header.nbands);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[i],"-swap"))
        swap=TRUE;
      else if(!strcmp(argv[i],"-yearcell"))
        header.order=YEARCELL;
      else
      {
        fprintf(stderr,"Error: invalid option '%s'.\n",argv[i]);
        fprintf(stderr,USAGE);
        return EXIT_FAILURE;
      }
    }
    else 
      break;
  if(argc<i+2)
  {
    fprintf(stderr,"Filenames missing.\n");
    fprintf(stderr,USAGE);
    return EXIT_FAILURE;
  }
  file=fopen(argv[i],"rb");
  if(file==NULL)
  {
    fprintf(stderr,"Error opening '%s': %s\n",argv[i],strerror(errno));
    return EXIT_FAILURE;
  }
  n=getfilesizep(file)/header.nyear/sizeof(Data);
  printf("Number of cells: %d\n",n);
  data=(Data *)malloc(n*sizeof(Data)*header.nyear);
  if(data==NULL)
  {
    fprintf(stderr,"Error allocating memory.\n");
    return EXIT_FAILURE;
  }
  if(fread(data,sizeof(Data),n*header.nyear,file)!=n*header.nyear)
  {
    fprintf(stderr,"Error reading data from '%s'.\n",argv[i]);
    return EXIT_FAILURE;
  }
  if(swap)
    for(k=0;k<n*header.nyear;k++)
      for(j=0;j<NMONTH;j++)
        data[k][j]=swapshort(data[k][j]); 
  fclose(file);
  file=fopen(argv[i+1],"wb");
  if(file==NULL)
  {
    fprintf(stderr,"Error creating '%s': %s\n",argv[i+1],strerror(errno));
    return EXIT_FAILURE;
  }
  header.ncell=n;
  fwriteheader(file,&header,LPJ_CLIMATE_HEADER,LPJ_CLIMATE_VERSION);
  if(header.order==CELLYEAR)
  {
    for(i=0;i<header.nyear;i++)
      for(j=0;j<n;j++)
        fwrite(data+j*header.nyear+i,sizeof(Data),1,file);
  }
  else
    fwrite(data,sizeof(Data),header.nyear*n,file);
  free(data);
  fclose(file);
  return EXIT_SUCCESS;
} /* of 'main' */
